cozy/codegen/rpython.py

```diff
@@ -0,0 +1,422 @@
+                    self.write(", ")
+
+    def visit_SMapDel(self, node):
+        with self.stmt():
+            self.write("del ")
+            self.visit(node.map)
+            self.write("[")
+            self.visit(node.key)
+            self.write("]")
+
+    def visit_SMapUpdate(self, node):
+        with self.stmt():
+            self.visit(node.val_var)
+            self.write(" = ")
+            self.visit(node.map)
+            self.write("[")
+            self.visit(node.key)
+            self.write("]")
+        self.visit(node.change)
+        # Writeback.
+        with self.stmt():
+            self.visit(node.map)
+            self.write("[")
+            self.visit(node.key)
+            self.write("] = ")
+            self.visit(node.val_var)
+
+    def visit_Op(self, node):
+        if node.assumptions:
+            import pdb; pdb.set_trace()
+        args = ", ".join([name for name, _ in node.args])
+        with self.python_block("def ", node.name, "(self, ", args, ")"):
+            # Since the body is a statement, we don't need to start the
+            # statement right here. Just go straight to the visitor. ~ C.
+            self.visit(node.body)
+            # And pass, in case nothing happened in the body.
+            self.write_stmt("pass")
+
+    def visit_Query(self, node):
+        if node.assumptions:
+            import pdb; pdb.set_trace()
+        args = ", ".join([name for name, _ in node.args])
+        with self.python_block("def ", node.name, "(self, ", args, ")"):
+            with self.stmt():
+                self.write("return ")
+                self.visit(node.ret)
+
+    def visit_Spec(self, spec, state_exps, sharing, abstract_state=()):
+        self.write_stmt("from rpython.rlib.objectmodel import specialize")
+
+        for func in spec.extern_funcs:
+            func_args = ", ".join([name for name, _ in func.args])
+            d = {n: n for n, _ in func.args}
+            with self.python_block("def ", func.name, "(", func_args, ")"):
+                # This is the style, so just go with it. ~ C.
+                body = func.body_string.format(**d)
+                self.write_stmt("return ", body)
+
+        with self.python_block("class ", spec.name, "(object)"):
+            init_args = ", ".join([name for name, _ in abstract_state])
+            with self.python_block("def __init__(self, ", init_args, ")"):
+                for name, exp in state_exps.items():
+                    self.instance_names.add(name)
+                    with self.stmt():
+                        self.write("self.", name, " = ")
+                        self.visit(exp)
+
+            for meth in spec.methods:
+                self.visit(meth)
+
+        for op, b in self.needs.items():
+            if not b:
+                continue
+            # Suddenly, globals()! These helpers are textually defined at the
+            # end of the module. ~ C.
+            self.write(globals()[op.upper()])
```

Comment:
This is fine, but I would slightly prefer to define `ARGMAX`, `ARGMIN`, etc. in a dictionary instead of using `globals()` like this.

(FWIW: What you've done here to convert Cozy AST nodes into helper functions is much cleaner than what the C++ or Java implementations currently do, and I heartily approve.)

Reply:
The helper functions could live in their own little RPython module. Maybe I could tack some Hypothesis tests onto that module too. We need a support module anyway, with classes for `Bag` and `Set`.

Doing the full codegen was on my mind, but also I didn't want to slog through debugging temporary variables needed for some of the inline functions. Cozy has nice expression-vs-statement semantics, but it means that codegen gets a little harder if I need to declare temporary names in an expression. So any time I found a case like that, I factored it out to a helper function.

As usual, `globals()` is terrible and I apologize. Friends don't let friends `globals()`.